[ONNX] Refactoring serialization of ONNX initializers to be name-based (Resubmission)

spandantiwari opened the pull request:
@houseroad - this is the resubmission of https://github.com/pytorch/pytorch/pull/17420, as suggested.

Commit:
Onnx export - refactoring to serialize ONNX initializers based on names and not order.

Commit:
Incorporate review feedback.

Commit:
Fixing how empty dict is passed in ONNX export.

Commit:
Fixing default argument style for unordered_map to fix test failures.

Commit:
Style fixes to assuage flake8.

Commit:
Updating two ONNX test expect files.

Commit:
Always writing initializers to ONNX in the same order.

Commit:
Updating export method to use std::map for determinstic order of serialization of initializers.

houseroad commented:
Apparently, this break several internal pipelines... I need more time to investigate what's going on.

houseroad reviewed (changes requested):
Put it on hold before the internal tests pass.

spandantiwari commented:
@houseroad - OK, let me know if I can help.

houseroad commented on torch/onnx/utils.py:
```diff
@@ -233,7 +233,11 @@ def _model_to_graph(model, args, f, verbose=False, training=False,
-        params = list(_unique_state_dict(model).values())
+        params = list(_unique_state_dict(model, keep_vars=True).values())
+
+    input_and_param_names = [val.uniqueName() for val in graph.inputs()]
```
Here we didn't use inputs provided by users to update our param_dict.

spandantiwari replied:
That's right. Fixed.

houseroad reviewed (commented):
The problem is only triggered when we set `input_names` and using non-protobuf serialization format (e.g., ZIP_ARCHIVE). We should add a test case to guard such situation.

Commit:
Fixed issue related to parameter renaming during ONNX export.

spandantiwari commented:
Fixed the issue with setting input_names and using non-protobuf serialization format (e.g., ZIP_ARCHIVE). Also added a test for guarding against that.

Commit:
Fixing lint issues.

Commit:
Updating expect files for tests whose initializer order has changed.

houseroad commented on torch/onnx/utils.py:
```diff
@@ -245,6 +245,10 @@ def _model_to_graph(model, args, f, verbose=False, training=False,
                 if i >= user_input_num:
                     inp.setUniqueName(param_names[i - user_input_num])
 
+    input_and_param_names = [val.uniqueName() for val in graph.inputs()]
```
let's move this part after `_optimize_graph`

Since `_optimize_graph` may change the name of the inputs (including initializers/parameters)

spandantiwari replied:
Done.

Commit:
Moving params_dict generation to after _optimize_graph().

houseroad commented on torch/onnx/utils.py:
```diff
@@ -247,18 +247,22 @@ def _model_to_graph(model, args, f, verbose=False, training=False,
 
     graph = _optimize_graph(graph, operator_export_type)
 
+    input_and_param_names = [val.uniqueName() for val in graph.inputs()]
```
Okay, after digging into the last problem, let's move these three new statements to the end of this method (i.e., `_model_to_graph`). Because we may also change the value names in _set_input_and_output_names. (Very rare case, when input_names conflicts with existing weights, and these weights won't be covered by input_names, and they will be renamed to something else, because their names are occupied by some other nodes covered by inputs)

spandantiwari replied:
Good point. Fixed.

houseroad commented on torch/onnx/utils.py:
```diff
@@ -247,18 +247,22 @@ def _model_to_graph(model, args, f, verbose=False, training=False,
             output.inferTypeFrom(tensor)
 
-    _set_input_and_output_names(graph, input_names, output_names)
+    _set_input_and_output_names(graph, input_names, output_names, params_dict)
```
if we move the creation of `params_dict` after this statement, then we don't need to update the params_dict in `_set_input_and_output_names` any more.

spandantiwari replied:
Removed the update logic in `_set_input_and_output_names`.

houseroad commented on torch/onnx/utils.py:
```diff
@@ -351,7 +355,7 @@ def _export(model, args, f, export_params=True, verbose=False, training=False,
 
 
-def _set_input_and_output_names(graph, input_names, output_names):
+def _set_input_and_output_names(graph, input_names, output_names, params_dict):
```
We don't need to change this function anymore.

spandantiwari replied:
Reverted.

houseroad reviewed (commented):
Let's add a test case to cover the input_names conflict with existing parameter names. 

Something like model has 2 parameters and 1 input.
export(model, input_names=['input1', 'p2']) and the second parameter's name is p2, and it will be renamed to p2.1 after exporting to onnx.

Commit:
Moving params_dict creation to the end of _model_to_graph.

spandantiwari commented:
Moved the `params_dict` creation code to the end as suggested. Added a new test point that checks tests the deliberate parameter name clobber.

spandantiwari commented:
@houseroad - the failures seem unrelated. I have seen some of them in previous runs as well.

houseroad commented on test/onnx/test_pytorch_onnx_caffe2.py:
```diff
@@ -207,6 +207,63 @@ def forward(self, input):
+
+        import io
+        f = io.BytesIO()
+        from torch.onnx import ExportTypes
```
Nit: lift this import to the beginning of the file.

spandantiwari replied:
Done.

houseroad commented on test/onnx/test_pytorch_onnx_caffe2.py:
```diff
@@ -207,6 +207,63 @@ def forward(self, input):
+                           input_names=['input1', 'fc1.bias'], _retain_param_name=False)
+
+        f.seek(0)
+        import caffe2.python.onnx.backend as c2
```
Nit: lift this import to the beginning of the file. So we just import once

spandantiwari replied:
Done.

houseroad commented on test/onnx/test_pytorch_onnx_caffe2.py:
```diff
@@ -207,6 +207,63 @@ def forward(self, input):
+                           input_names=['input1', 'parameter1', 'parameter2'])
+
+        f.seek(0)
+        import caffe2.python.onnx.backend as c2
```
Nit: lift this import to the beginning of the file.

spandantiwari replied:
This import is already there at top. Done.

houseroad commented on test/onnx/test_pytorch_onnx_caffe2.py:
```diff
@@ -207,6 +207,63 @@ def forward(self, input):
+
+        import io
+        f = io.BytesIO()
+        from torch.onnx import ExportTypes
```
Nit: lift this import to the beginning of the file.

spandantiwari replied:
Done.

houseroad commented on torch/csrc/jit/python_ir.cpp:
```diff
@@ -14,6 +14,7 @@
 
 #include <iostream>
 #include <sstream>
+#include <unordered_map>
```
remove?

spandantiwari replied:
Done.

houseroad reviewed (approved):
Looks good! Finally, we fixed all the problems.

Left some nits.

Commit:
Cleaning up import statements.

spandantiwari commented:
@houseroad - Great to hear that all tests are passing. Thanks for your tremendous help!
I have updated the PR based on the feedback.
